### ISSUE TITLE
Performance decrease on reused OIDC Access Tokens

### ISSUE THREAD

**Issue by pellmont** [OpenLiberty/open-liberty#15379]
Re-Use of an Access Token for multiple REST Requests will lead to performance decrease.

Reason: The Subject is growing with each request, because a new Hashtable containing the access token is added to the cached subject. This is especially significant if you use larger access tokens (like JWT Tokens).

Suspected source of trouble:
WebProviderAuthenticationHelper in com.ibm.ws.webcontainer.security
Method removeSecurityNameAndUniquedIdFromHashtable.
In this method an attempt to remove two Properties from the Hashtable leads to addition of an extra Hashtable per request (since the original Hashtable is not on the Subject from the auth cache!).

Possible solution: Only re-add the Hashtable if it could be removed before:
```
  private void removeSecurityNameAndUniquedIdFromHashtable(Subject subject, Hashtable<String, ?> props, boolean mapIdentityToRegistryUser)
  {
    if ((!mapIdentityToRegistryUser) && (!subject.isReadOnly()))
    {
      Set<Object> privateCredentials = subject.getPrivateCredentials();
      if (privateCredentials.remove(props)) {
        props.remove("com.ibm.wsspi.security.cred.uniqueId");
        props.remove("com.ibm.wsspi.security.cred.securityName");
        if (!props.isEmpty()) {
          privateCredentials.add(props);
        }
      }
    }
  }

```

**Comment by una-tapa** [comment]
@mbroz2 Thanks!!

**Comment by pellmont** [comment]
Thanks for caring. Let me know if you need further information or if I can help in any way...

**Comment by una-tapa** [comment]
I am going to implement suggested code to see if it goes thru the automated test so it will be easier to consult with the SSO team when they come back from the holiday break.

**Comment by pellmont** [comment]
Is there any news on this issue? Can I help in some way to progress?

**Comment by una-tapa** [comment]
@jhanders34 @ayoho  Is there any way to get this reviewed soon ?  @pellmont 's code looks good, it got the green build... hopefully it can be merged soon..

**Comment by una-tapa** [comment]
Adam suggested @arkarkala or @utle 's review.  Thanks for patience > @pellmont

**Comment by una-tapa** [comment]
@pellmont  

My sincere apology that it is taking long time to get the code reviewed. 
While waiting, I would like to offer a jar that implemented your code so you don't experience the performance degradation.  
In that case, I would like to find out Liberty version you are running on.  Also if you allow me, I would like to put some obvious debug output to show the architects that your suggested fix is useful.  (Hopefully!)

**Comment by una-tapa** [comment]
@utle suggested to ask for trace. 
com.ibm.ws.security.*=all:com.ibm.webcontainer.security.*=all 
I will post "how to trace" info.

**Comment by pellmont** [comment]
Hi @una-tapa 

Up to now we used 20.0.0.12, but I just updated to 21.0.0.1  (the effect is the same), so you could build the jar against the most recent version.
let me know where to put the jar and how to enable the tracing...

**Comment by una-tapa** [comment]
Hi, @pellmont 
My apology it took me a while.  
Our developer would like to review the trace.  Kindly provide us at your convenience. 
For details how to get a trace, please take a look at  https://github.com/una-tapa/pellmont-file
Thank you!

**Comment by pellmont** [comment]
Hi @una-tapa 

Now it's my turn to apologize for being late.
I just managed to run the tests with the JAR you prepared for me. With the new JAR the performance decrease doesn't show up anymore. And what is slightly surprising: no matter whether the flag in jvm.properties is true or false.
As for the traces: you just want to see the SystemOut entries in messages.log? Or should there be any other logfiles around?

**Comment by una-tapa** [comment]
Hi, @pellmont  

pellmont >> With the new JAR the performance decrease doesn't show up anymore

Glad to hear from you.  Thanks for the great news!   

I feel this test would be sufficient to prove the code is effective, but let me try to provide (hopefully) a better steps to produce the trace.  
- Please replace my test jar with the original `com.ibm.ws.webcontainer.security_1.0.48.jar` to reproduce the issue. 
- Please place  [bootstrap.properties](https://github.com/una-tapa/pellmont-file/blob/master/bootstrap.properties) in the same directory where you find  `server.xml`
- Once the server picks up the trace option, trace.log should be created in the same directory where the `messages.log` is. 
- Run the scenario to capture the performance degradation in the trace. 
- Please share the trace with us. 
 
Thanks again.  Hiroko

SUMMARY (for my consultation later with the team)  

This git issue is a bit unfortunate... it has been sitting more than 3 months for various reasons. 
A while ago, I sent pellmont a jar file with his suggested code (https://github.com/OpenLiberty/open-liberty/pull/15392)  .  According to his test above, the performance degradation no longer happens. 
To recap,  pellmont's suggestion is to prevent the code from automatically adding tokens (It will help with performance especially large tokens like JWT)   Although we have not received the trace yet,  I personally think this is really good code. If there is a concern about breaking[ the zero migration policy](https://www.ibm.com/support/pages/websphere-migration-knowledge-collection-migrating-liberty#:~:text=The%20Liberty%20zero%2Dmigration%20architecture,binaries%20comes%20with%20no%20surprises.), I would like to check in his suggested code with a switch.

**Comment by pellmont** [comment]
Hi @una-tapa 

I finally managed to produce the tracelog (the bootstrap.properties had ben overwritten...)
You can find the tracelog here: https://github.com/pellmont/tracelogs-openliberty-performance-issue

**Comment by una-tapa** [comment]
Hi @pellmont 

Thank you so much! 

Now I see a lot of tracing errors (instead of catching the actual performance issue)  :-(   

The trace spec at the top of the log looks great to me. 
trace.specification = *=info:com.ibm.ws.security.*=all:com.ibm.ws.webcontainer.*=all:com.ibm.wsspi.security.*=all:com.ibm.wsspi.webcontainer.*=all

I am wondering if my bootstrap caused this trouble.... if so please remove the bootstrap. It is for customers who do not edit files often but in your case, you may have the trace configured in the server.xml already?

**Comment by pellmont** [comment]
Hi @una-tapa 

I didn't have a trace setting, but another setting which caused the problems. I updated the tracelog in the repository... Hope this helps

Best regards
Pascal

**Comment by una-tapa** [comment]
@arkarkala @utle 

We just discussed this issue in today's scrum.  Pascal provided the trace here.  https://github.com/pellmont/tracelogs-openliberty-performance-issue

The trace was taken while my patch was enabled.  (<=Pascal's idea)  Patched code is from line 141 to 167.  
https://github.com/una-tapa/pellmont-file/blob/master/WebProviderAuthenticatorHelper.java#L141 

If the code path were the current code, Liberty would have added the access token 210 more times. 
Please review, and hopefully approve the PR.  
https://github.com/OpenLiberty/open-liberty/pull/15392/files

Thanks! 

--- Subject from the trace -- 
Subject:
	Principal: WSPrincipal:p9786
	Public Credential: com.ibm.ws.security.credentials.wscred.WSCredentialImpl@693efadf,realmName=https://oidc-dev.css.ch/auth/realms/css,securityName=p9786,realmSecurityName=https://oidc-dev.css.ch/auth/realms/css/p9786,uniqueSecurityName=p9786,primaryGroupId=null,accessId=user:https://oidc-dev.css.ch/auth/realms/css/p9786,groupIds=[group:https://oidc-dev.css.ch/auth/realms/css/g_cam_admin, group:https://oidc-dev.css.ch/auth/realms/css/g_ppva_kube, group:https://oidc-dev.css.ch/auth/realms/css/g_was_consoper, group:https://oidc-dev.css.ch/auth/realms/css/g_was_blob_user, group:https://oidc-dev.css.ch/auth/realms/css/offline_access, group:https://oidc-dev.css.ch/auth/realms/css/uma_authorization]
	Private Credential: JsonWebToken:{"jti":"9ac969f8-2512-40ef-804e-4a62e9e80e8f","exp":1616596663,"nbf":0,"iat":1616596363,"iss":"https://oidc-dev.css.ch/auth/realms/css","aud":"account","sub":"0ca74ef2-24cf-485a-adf9-44569031c284","typ":"Bearer","azp":"bsfw-sample","auth_time":0,"session_state":"a55002eb-50fb-431f-9d84-a5fcb3201862","acr":"1","resource_access":{"account":{"roles":["manage-account","manage-account-links","view-profile"]}},"scope":"profile realmname roles email","realm":"css","preferred_username":"p9786","email":"account2.testusertechnischearchi@css.ch","realmroles":["g_cam_admin","g_ppva_kube","g_was_consoper","g_was_blob_user","offline_access","uma_authorization"]}
	Private Credential: com.ibm.ws.security.token.internal.SingleSignonTokenImpl@27032b
	Private Credential: {com.ibm.wsspi.security.cred.cacheKey=1861815372, com.ibm.wsspi.security.cred.realm=https://oidc-dev.css.ch/auth/realms/css, com.ibm.wssi.security.oidc.client.credential.storing.utc.time.milliseconds=1616596369606, com.ibm.ws.authentication.internal.assertion=true, com.ibm.wsspi.security.cred.groups=[group:https://oidc-dev.css.ch/auth/realms/css/g_cam_admin, group:https://oidc-dev.css.ch/auth/realms/css/g_ppva_kube, group:https://oidc-dev.css.ch/auth/realms/css/g_was_consoper, group:https://oidc-dev.css.ch/auth/realms/css/g_was_blob_user, group:https://oidc-dev.css.ch/auth/realms/css/offline_access, group:https://oidc-dev.css.ch/auth/realms/css/uma_authorization], **access_token=eyJhbGciOiJS ... (VERY HUGE TOKEN) .... ohHy3wFysPEMesgwa05rKUxLz4MEj00nC9Q}**

-------------------------------------------------------------

**Comment by una-tapa** [comment]
@pellmont Thanks for contributing the code!  The issue is closed after merge.

**Comment by utle** [comment]
Fixed by https://github.com/OpenLiberty/open-liberty/pull/17147

**Comment by una-tapa** [comment]
This Issue 15379 is fixed by PR 15392 (https://github.com/OpenLiberty/open-liberty/pull/15392) 
Ut's PR 17147 fixes Issue17155

Just so the admins will get the list of issues correctly. 
cc: @utle
